crates/bpe/src/byte_pair_encoding.rs

```diff
@@ -552,3 +552,74 @@ impl BytePairEncoding {
         encoded
     }
 }
+
+#[cfg(feature = "rand")]
+fn is_char_boundary(b: u8) -> bool {
+    // Single byte encodings satisfy the bit pattern 0xxxxxxx, i.e. b < 128
+    // Continuation bytes satisfy the bit pattern 10xxxxxx, i.e. b < 192
+    // The rest are bytes belonging to the first byte of multi byte encodings (11xxxxxx): b >= 192
+    // When interpreting the byte representation as signed integers, then numbers in the range 128..192
+    // correspond to the smallest representable numbers. I.e. the two ranges [0, 128) and [192, 256) can
+    // be tested with a single signed comparison.
+    b as i8 >= -0x40 // NB: b < 128 || b >= 192
+}
+
+#[cfg(feature = "rand")]
+pub fn create_test_string(bpe: &BytePairEncoding, min_bytes: usize) -> String {
+    use rand::{thread_rng, Rng};
+    // the bytes we accumulated thus far
+    let mut bytes = Vec::new();
+    // the tokens we added so we can backtrack
+    let mut tokens = Vec::new();
+    // the number of valid UTF-8 bytes
+    let mut valid_bytes = 0;
+    'keep: while valid_bytes < min_bytes {
+        // try a few times to find a suitable token
+        for _ in 0..8 {
+            // pick a random token and provisionally add it
+            let i = thread_rng().gen_range(0..bpe.num_tokens());
+            bytes.extend(bpe.token_bytes(i as u32));
+            // test if the additional bytes are valid utf-8
+            // the last character is not included, because it may be incomplete
+            let last = bytes
+                .iter()
+                .rev()
+                .find_position(|b| is_char_boundary(**b))
+                .map_or(0, |(offset, _)| bytes.len() - (offset + 1));
+            assert!(last >= valid_bytes);
+            if std::str::from_utf8(&bytes[valid_bytes..last]).is_ok() {
+                tokens.push(i);
+                valid_bytes = last;
+                continue 'keep;
+            } else {
+                bytes.truncate(bytes.len() - bpe.token_len(i as u32));
+            }
+        }
```

Comment:
First time actually looking at this code :)
I'm not so sure that this procedure has any reasonable chance of constructing valid utf8 from "broken" tokens.
Essentially there are VERY few of those tokens and the chance of picking the correct one is extremely low.
The original idea of picking tokens was that the chance of picking longer interesting token sequences is higher.
But, this doesn't really work with broken utf8 sequences so easily. You would have to store valid token pairs. But there are just too many pairs that this is computationally reasonable.

Reply:
I see. I did a quick test and for both cl100k and o200k around 99% of the tokens are valid utf-8. o200k has more tokens containing multi-byte characters. For cl100k, almost 95% match the old condition `token.iter().all(|b| is_char_boundary(*b))`, but only 64% of o200k tokens does.

I guess I can simplify this function a bit by only testing for valid utf-8 tokens.